servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,547 @@
+                        pendingDemand = emittedCount = prevPendingDemand;
+                        signals.clear();
+                    }
+                    Object t;
+                    while (emittedCount < prevPendingDemand && (t = signals.poll()) != null) {
+                        try {
+                            if (sendToTarget(t)) {
+                                ++emittedCount;
+                            }
+                        } catch (Throwable cause) {
+                            delayedCause = catchUnexpected(delayedCause, cause);
+                        }
+                    }
+
+                    // check if a terminal event is pending, or give back demand.
+                    if (emittedCount == prevPendingDemand) {
+                        for (;;) {
+                            try {
+                                t = signals.peek();
+                                if (t == MAPPED_SOURCE_COMPLETE) {
+                                    signals.poll();
+                                    tryRequestMoreFromUpstream();
+                                } else if (t instanceof FlatMapPublisherSubscriber) {
+                                    signals.poll();
+                                    ((FlatMapPublisherSubscriber<?, ?>) t).replenishDemandAndClearSignalsQueued();
+                                } else {
+                                    break;
+                                }
+                            } catch (Throwable cause) {
+                                delayedCause = catchUnexpected(delayedCause, cause);
+                            }
+                        }
+
+                        if (t instanceof TerminalNotification) {
+                            sendToTarget((TerminalNotification) t); // if this throws its OK as we have terminated
+                        } else {
+                            t = terminalNotification;
+                            if (t != null && t != complete()) { // don't wait for demand to process an error
+                                sendToTarget((TerminalNotification) t); // if this throws its OK as we have terminated
+                            }
+                        }
+                    } else {
+                        assert emittedCount < prevPendingDemand;
+                        pendingDemandUpdater.accumulateAndGet(this, prevPendingDemand - emittedCount,
+                                FlowControlUtils::addWithOverflowProtection);
+                    }
+                } finally {
+                    tryAcquire = !releaseLock(emittingLockUpdater, this);
+                }
+            }
+
+            if (delayedCause != null) {
+                throwException(delayedCause);
+            }
+        }
+
+        private void tryRequestMoreFromUpstream() {
+            if (--upstreamDemand == 0) { // heuristic to replenish demand when it is exhausted.
```

Comment:
Good question. This is not intentional and the motivation for this heuristic was to batch demand to avoid `request(1)` when each mapped source completes. I'll switch to batching demand within the draining of the signal delivery queue. Unless there is concurrency on completion this will digress to `request(1)` but we can enhance later with something more sophisticated batching (e.g. configurable heuristic).